include/zim/archive.h

```diff
@@ -0,0 +1,361 @@
+       *
+       * @tparam order The order to use to iterate the archive.
+       */
+       template<EntryOrder order>
+      iterator<order> end() const
+        { return iterator<order>(m_impl, entry_index_type(getEntryCount())); }
+
+      /** Find the position of an entry using it path.
+       *
+       * @return A iterator to the entry corresponding to path.
+       *         `end` iterator if no entry correspond the the path.
+       */
+      iterator<EntryOrder::pathOrder>  findByPath(const std::string& path) const;
+
+      /** Find the position of an entry using it title.
+       *
+       * @return A iterator to the entry corresponding to title.
+       *         `end` iterator if no entry correspond the the title.
+       */
+      iterator<EntryOrder::titleOrder> findByTitle(const std::string& title) const;
+
+      /** hasChecksum.
+       *
+       * The checksum is not the checksum of the file.
+       * It is an internal checksum stored in the zim file.
+       *
+       * @return True if the archive has a checksum.
+       */
+      bool hasChecksum() const;
+
+      /** getChecksum.
+       *
+       * @return the checksum stored in the archive.
+       *         If the archive has no checksum return an empty string.
+       */
+      std::string getChecksum() const;
+
+      /** Check that the zim file is valid (in regard to its checksum).
+       *
+       *  If the zim file has no checksum return false.
+       *
+       *  @return True if the file is valid.
+       */
+      bool check() const;
+
+      /** Check if the file is split in the filesystem.
+       *
+       *  @return True if the archive is split in different file (foo.zimaa, foo.zimbb).
+       */
+      bool is_multiPart() const;
+
+#ifdef ZIM_PRIVATE
+      cluster_index_type getClusterCount() const;
+      offset_type getClusterOffset(cluster_index_type idx) const;
+      entry_index_type getMainEntryIndex() const;
+#endif
+
+    private:
+      std::shared_ptr<FileImpl> m_impl;
+  };
+
+  template<EntryOrder order>
+  entry_index_type _toPathOrder(const FileImpl& file, entry_index_type idx);
+
+  template<>
+  entry_index_type _toPathOrder<EntryOrder::pathOrder>(const FileImpl& file, entry_index_type idx);
+  template<>
+  entry_index_type _toPathOrder<EntryOrder::titleOrder>(const FileImpl& file, entry_index_type idx);
+  template<>
+  entry_index_type _toPathOrder<EntryOrder::efficientOrder>(const FileImpl& file, entry_index_type idx);
+
+  template<EntryOrder order>
+  class Archive::iterator : public std::iterator<std::bidirectional_iterator_tag, Entry>
+  {
+    public:
+      explicit iterator(const std::shared_ptr<FileImpl> file, entry_index_type idx)
+        : m_file(file),
+          m_idx(idx),
+          m_entry(nullptr)
+      {}
+
+      iterator(const iterator<order>& other)
+        : m_file(other.m_file),
+          m_idx(other.m_idx),
+          m_entry(new Entry(*other.m_entry))
+      {}
```

Comment:
Because the ownership is not shared. The iterator is the only owner of the entry.
I don't think we use a lot the copy constructor in the classical pattern. And construct a `shared_ptr` is more expensive than a `unique_ptr`. So it is probably better to do a deep copy (cheap anyway) time to time than construct a `shared_ptr` all the time.

In fist instance, I use a pointer here only because we cannot have an invalid entry. So I use a pointer to be able to have a "invalid" (null) pointer.

Reply:
Just a little piece of useful information - `shared_ptr` is cheaper to construct via [`std::make_shared`](https://en.cppreference.com/w/cpp/memory/shared_ptr/make_shared):

> ### Notes
> * `std::shared_ptr<T>(new T(args...))` performs at least two allocations (one for the object `T` and one for the control block of the shared pointer), while `std::make_shared<T>` typically performs only one allocation (the standard recommends, but does not require this; all known implementations do this)